src/module.c

```diff
@@ -633,14 +634,15 @@ void moduleFreeContext(RedisModuleCtx *ctx) {
 /* Create a module ctx and keep track of the nesting level.
  *
  * Note: When creating ctx for threads (RM_GetThreadSafeContext and
- * RM_GetDetachedThreadSafeContext) one needs to pass module=NULL
- * because we only need to track nesting level in the main thread
+ * RM_GetDetachedThreadSafeContext) we do not bump up the nesting level
+ * because we only need to track of nesting level in the main thread
  * (only the main thread uses propagatePendingCommands) */
-void moduleCreateContext(RedisModuleCtx *out_ctx, RedisModule *module) {
+void moduleCreateContext(RedisModuleCtx *out_ctx, RedisModule *module, int ctx_flags) {
     memset(out_ctx, 0 ,sizeof(RedisModuleCtx));
     out_ctx->getapifuncptr = (void*)(unsigned long)&RM_GetApi;
-    if (module) {
-        out_ctx->module = module;
+    out_ctx->module = module;
+    out_ctx->flags = ctx_flags;
+    if (out_ctx->module && !(ctx_flags & REDISMODULE_CTX_THREAD_SAFE)) {
```

Comment:
@MeirShpilraien please put a third set of eyes on this to make sure we're not overlooking anything.